collection-scripts/gather_crio_profiling

```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+
+# safeguards
+set -o nounset
+set -o errexit
+set -o pipefail
+
+# TODO: check which is the right "BASE_COLLECTION_PATH" to use
+BASE_COLLECTION_PATH="must-gather"
+CRIO_LOG_PATH="${BASE_COLLECTION_PATH}/pprof/crio"
+CRIO_PROFILING_DURATION_SEC=30
+
+mkdir -p "$CRIO_LOG_PATH"
+
+# NOTE: followed gather_core_dumps script
+
+_gather_crio_profiling_data_from_node() {
+	local node="$1"
+    local tmpdir="/tmp/crio_pprof"
+    local curl_pprof_ep="curl --unix-socket /host/var/run/crio/crio.sock http://localhost/debug/pprof/"
+
+    local cp_timeout=$((CRIO_PROFILING_DURATION_SEC+120)) # debug pod timeout waiting for data to be copied
+    local cp_retry_num=5 # how many times we will try to copy the data if we fail
+    local cp_retry_interval=10 # how many seconds to wait before retrying copy data
+
+    # Get debug pod's name
+	debugPod=$(oc debug --to-namespace=default node/${node} -o jsonpath='{.metadata.name}')
+    if [ -z "$debugPod" ]
+    then
+        echo "ERROR: cannot start debug pod for node ${node}"
+        return
+    fi
+	echo "INFO: starting CRI-O profiling on node ${node}"
+    # Start debug pod collecting CRI-O profiling data in the "default" namespace and let it wait we
+    # copy from the pod the profiling data.
+    oc debug --to-namespace="default" node/${node} -- /bin/bash -c "\
+        mkdir -p ${tmpdir} \
+        && ${curl_pprof_ep}profile?seconds=${CRIO_PROFILING_DURATION_SEC} > ${tmpdir}/prof.out \
+        && ${curl_pprof_ep}heap > ${tmpdir}/heap.out \
+        && sleep ${cp_timeout} \
+        " > /dev/null 2>&1 &
+
+    # Data will be ready only after profiling is over.
+    sleep ${CRIO_PROFILING_DURATION_SEC}
```

Comment:
Part of the reason why I find this fragile is that when we 'collect' data we have to 'sleep' (which delays or stalls the overall collection time), and we can't get that back (its just lots time). 

Consider the following:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: sos-collector
  labels:
    name: sosreport-container
spec:
  hostNetwork: true
  nodeName: "NODE NAME"   ## TODO: Template or Change
  hostPID: true
  restartPolicy: Never
  priorityClassName: "system-cluster-critical"
  tolerations:
  - key: node-role.kubernetes.io/master
    operator: "Equal"
    effect: "NoExecute"
    tolerationSeconds: 3600
  - key: node-role.kubernetes.io/master
    operator: "Equal"
    effect: "NoSchedule"
  containers:
  - name: sos-extractor
    image: registry.redhat.io/rhel8/support-tools
    command: ["/bin/bash", "-c", "trap : TERM INT; sleep infinity & wait"]
    volumeMounts:
    - mountPath: /tmp/sos-data
      name: sos-data
  initContainers:
  - name: sos-collector
    image: registry.redhat.io/rhel8/support-tools
    command: ["sosreport", "-s", "/host", "--tmp-dir", "/tmp/sos-data", "-c", "always", "--batch", "-v"]
    securityContext:
      runAsUser: 0
      privileged: True
    volumeMounts:
    - mountPath: /host
      name: host-volume
    - mountPath: /tmp/sos-data
      name: sos-data
  volumes:
  - name: host-volume
    hostPath:
      path: /
      type: Directory
  - name: sos-data
    emptyDir: {}
```
The cli execution then looks like the following.

```
oc create -f sos_pod.yaml
oc wait --for=condition=Ready pod/sos-collector --timeout 10m && oc cp -c sos-extractor sos-collector:/tmp/sos-data/ .
```

What this does is it 'starts a collection container' and only if that completes it 'starts an exfiltration container' that the oc wait (can pick up the signal - start of the pod) and oc cp/rsync command can continue driving the script. 

This allows us a slightly more uniform way to 'start a collection' and 'gather the collection data' all while avoiding sleeps and other unnecessary waiting or unavoidable time sucks.

Reply:
Well, this is an interesting approach I'm evaluating, thanks for sharing (I learned something!).
Anyway, I don't see a functional advantage to switch to that.
While we use the "sleep" command in the PR to wait for the profiling job to end, we are doing it in a careful way (polling): we slip first _CRIO_PROFILING_DURATION_SEC_ (30 sec), as it is the time the profiling data will run (so we know the job cannot end up earlier) and than we pol the job for completion every _cp_retry_interval_ (10 sec.). We poll for a maximum of _cp_retry_num_ (5) times.
So, worst case if a profiling job succeed is a waste of 10 seconds while waiting.
worst case if a profiling job fails or get stuck is we will wait for 50 sec. (the succeding jobs running in parallel will last 30 seconds at the very minimum anyway).
The only blocking time is that we wait for all the running jobs to end before moving on.
So, if one (or more) debug pod gets stuck we will wait 50 secs. maximum. before moving on to the next must-gather task.
Seems to me the approach presented here cannot do better. What would be the advantage of it?

Reply:
Well, the sleep command is only used once to do the polling in the gather_node_util script now.